### PR TITLE
Change docker tag logic

### LINKED ISSUE #15622: Docker: re-evaluate tag heuristic (`:latest`, `:dev`)
Given the recent discussions in the discord `docker` channel again about the latest tag, I would like to bring the topic of a possible change of the `:latest` tag up.

New users frequently enter the rabbit hole of starting with `:latest` and finding themselves on a non-release build.

Given that transitioning downwards is not straightforward due to a required DB schema downgrade, I think it could make things easier for many people if `:latest` would equal `:1` and point to the latest release.

A dedicated `:dev` tag could make all people happy that want to be on the latest git revision.

An additional idea could be to make the `:1` tag somewhat more stable than `:latest` by only bumping to a new minor release after the second patch release.
I could think of some users who want more stability to appreciate such an option.

## Summary

- `:latest` points to the latest release
- `:1` or/and `:stable` point to the latest release if there are more than 2 patch releases available (e.g. 1.13.2, 1.14.2)
- `:dev` points to the latest git revision

### PR DESCRIPTION
**Pull request by pat-s**
fixes #15622

- [x] `:latest` -> `:dev`
- [ ] `:latest` = `:1` = `:stable`
- [ ] Update docs

@techknowlogick @sapk 

I could not find where the `:1` tag is defined on releases. Is it only triggered by a local heuristic on release?

This change should change HEAD version tag to `:dev`, unless I overlooked something.

Feedback/help welcome :)

### CODE REVIEW

**Comment by lafriks** [comment]
I don't like changing the meaning of tag. I would probably than prefer dropping it if we go this way

**Comment by 6543** [comment]
well we need **consensus** ... and more people testing gitea - especially **windows users** !

**Comment by techknowlogick** [comment]
> I could not find where the :1 tag is defined on releases. Is it only triggered by a local heuristic on release?

`:1` comes from the manifest plugin, the logic is if there is a semver-like tag, then it'll split it up into 1, 1.4, 1.4.0 (assuming a non-RC/alpha tag), if it is an RC/alpha tag then it'll just use the exact tag (found here: https://github.com/drone-plugins/drone-manifest/blob/master/tagging/tagging.go#L21-L56)

> I don't like changing the meaning of tag. I would probably than prefer dropping it if we go this way

the only meaning we are changing is `latest`, and that can't exactly be dropped.

**Comment by pat-s** [comment]
> and more people testing gitea - especially windows users !

That would be great ofc, but these people should be "testers" because they want to, not because they accidentally used `:latest` and then find themselves in a DB downgrading problem.
This happen(ed) quite often and is really a pain for everyone (users and contributors helping to downgrade).

> :1 comes from the manifest plugin, the logic is if there is a semver-like tag, then it'll split it up into 1, 1.4, 1.4.0 (assuming a non-RC/alpha tag), if it is an RC/alpha tag then it'll just use the exact tag (found here: https://github.com/drone-plugins/drone-manifest/blob/master/tagging/tagging.go#L21-L56)

Thanks, I will have a look how I can integrate `:latest` and `:stable` into it.

**Comment by 6543** [comment]
@pat-s ... good point

We can switch this schema each time migration level is equal on main and release branch ... so on each majour release

**Comment by lafriks** [comment]
Problem is that we will pretty much break all instances with this change using :latest as they won't start anymore as essentially we will force downgrade on them. Also in all recommendations it is warned not to use `:latest` because of multiple reasons (cache, best practices etc)

**Comment by techknowlogick** [comment]
@lafriks we'd wait until next major release to switch the `:latest` label, so that no downgrades would be possible.

**Comment by lafriks** [comment]
I would propose `:latest` tag builds show `This is development Gitea version. It's not recommended to be used on production environment. More info` on top of every page as warning (like one on try.gitea.io). Where on more info would display popup on what version be used and how to disable this warning by adding `HIDE_DEVELOPMENT_WARNING=true` in app.ini

**Comment by pat-s** [comment]
> I would propose :latest tag builds show This is development Gitea version. It's not recommended to be used on production environment. More info on top of every page as warning (like one on try.gitea.io). Where on more info would display popup on what version be used and how to disable this warning by adding HIDE_DEVELOPMENT_WARNING=true in app.ini

This could be done in addition to the new `:dev` tag. However I guess people opting in actively are probably just annoyed by this without a real benefit.

The issue that people end up unwanted in the dev build can IMO only be solved by switching the tag logic away from `:latest`.

**Comment by lafriks** [comment]
My proposal would be to drop `:latest`

**Comment by pat-s** [comment]
We could also get rid of `:latest` to have a more tidy tag setup. I don't have a strong opinion on this. However, I think that this would cause more disruption than refocusing to the stable releases because people would not get any updates anymore starting the next day (assuming auto-updates). 

If we do this, we want to communicate this upfront throughout the period of a point release?

**Comment by techknowlogick** [comment]
I have a strong opinion against getting rid of `:latest` because it is the default tag when pulling an image, and I believe it would cause more confusion than switching to stable.

**Comment by lafriks** [comment]
At least it would force to specify one 🤣

**Comment by pat-s** [comment]
@techknowlogick I've reduced the build matrix in this PR to only run the docker dry run pipelines temporarily.
I've also added a dry run for the manifest runs.

However, the manifest runs only get triggered by a tag / need a tag for correct image tagging.
I cannot push tags and was wondering if you/someone else could trigger a test run with a fictional tag (e.g. 99.99.99) to check if everything works? Not sure if this is possible in a PR from a forked repo though.

Maybe there is an easier way to test this which I am currently overlooking?

Also I think the `:latest` -> `:dev` should already be fine and the release tagging is still WIP.

**Comment by pat-s** [comment]
Maybe @zeripath can also shine light on this? 🙂

**Comment by zeripath** [comment]
> Maybe @zeripath can also shine light on this?

OK, I think we need to balance the benefits and risks of our current system in toto:

1. We have very few people who build gitea and test gitea latest deliberately. It's highly likely that the vast majority of our bug reports regarding breaks in main come from docker users who - unwittingly or not - are running latest and don't realise that this is unreleased.
2. Changing latest to match the last released version would fix that confusion but it would seriously slow down testing because of the loss of this otherwise large testing pool.
3. This would mean that migration bugs would only be spotted on release - which would mean that more people would be affected by them.
  - However, we now have migration tests and our dumping and restore tool is better - so that's somewhat ameliorated.
4. It would mean that 1.x.0 releases would have more bugs in them at the start.
  - This would mean that developer load for dealing with release bugs could be worse - i.e. on release day we're going to have to be prepared to fix a flurry of bugs in a quick 1.x.1 release. (Which we already face.)
  - And we may have to spend more time backporting.  (Looking at `git rev-list --count origin/release/v1.x...v1.1x.0` reveals that we currently have 119 backports on to 1.14 before we've even released 1.15. By the end of 1.13 we had 130, 1.12 was 112, 1.11 was 124 and 1.10 was 66. So our backport load is already quite high and increasing.)
5. Conversely the lack of apparent feature release would lead to more interest in getting a release out of the door.
  - Perhaps this would lead to more people trying the RCs as they'd be more important to them?

In the end though I'm tired of being flamed by (often novice) docker users - whilst I don't think we're doing anything wrong and it wasn't wrong originally - I think consensus has moved underneath us and most people on dockerhub have moved to ":latest" to represent the latest stable.

So unfortunately I think we probably do need to do it - but in doing so we're going to have acknowledge that there will be consequences on our ability to get things tested and likely we're going to have to change the way we manage releases and merging and in particular refactors.

**Comment by 6543** [comment]
If we go this way, we should add a new tag "main" who is the new master and realy encurage users to run this if they like to support us!

**Comment by 6543** [comment]
"dev" or "testung" sounds to unstable ...

**Comment by zeripath** [comment]
main then

**Comment by techknowlogick** [comment]
as much as I love to yak shave (I don't), we need to denote some level of "this isn't the stable release" so `main` could impart that it is the "main" image and it should be used over some other image. Images such as minio use `edge` for that, but I find dev/develop just as suitable.

**Comment by pat-s** [comment]
I'd also favor `dev` over main for the reasons outlined by @techknowlogick 

However I would still need some feedback how I/we can test drive this. See my last comment 🙂

**Comment by 6543** [comment]
if we go this direction I think it's time to make this pull ready

File: docker/manifest.rootless.tmpl [hunk 1]
```diff
@@ -1,9 +1,10 @@
-image: gitea/gitea:{{#if build.tag}}{{trimPrefix "v" build.tag}}{{else}}latest{{/if}}-rootless
+image: gitea/gitea:{{#if build.tag}}{{trimPrefix "v" build.tag}}{{else}}dev{{/if}}-rootless
 {{#if build.tags}}
 tags:
 {{#each build.tags}}
   - {{this}}-rootless
 {{/each}}
+  - "latest"
```
Comment (techknowlogick) [inline docker/manifest.rootless.tmpl]:
```suggestion
  - "latest-rootless"
```

**Comment by techknowlogick** [comment]
Closing in lieu of https://github.com/go-gitea/gitea/pull/16421

**Pull request closed without merging**